src/engine.ml

```diff
@@ -376,84 +384,77 @@ let merge_server_config_into_client config tls_data =
-            in
-            CHACHA20_POLY1305
-              {
-                my_key = Mirage_crypto.Chacha20.of_secret my_key;
-                my_implicit_iv;
-                their_key = Mirage_crypto.Chacha20.of_secret their_key;
-                their_implicit_iv;
-              }
-      in
-      let keys_ctx = { my_packet_id = 1l; their_packet_id = 1l; keys } in
-      (tls_data, keys_ctx)
+let get_tls_data ?with_premaster data =
+  let open Result.Infix in
+  Log.debug (fun m -> m "received tls payload %d bytes" (Cstruct.length data));
+  Packet.decode_tls_data ?with_premaster data >>| fun tls_data ->
+  Log.debug (fun m -> m "Received tls-data %a" Packet.pp_tls_data tls_data);
+  tls_data
+
+let kdf session cipher hmac_algorithm key tls_data =
+  let keys = derive_keys session key tls_data in
+  let maybe_swap (a, b, c, d) =
+    if Cstruct.(equal empty key.State.pre_master) then (c, d, a, b)
+    else (a, b, c, d)
+  in
+  let extract klen hlen =
+    ( Cstruct.sub keys 0 klen,
+      Cstruct.sub keys 64 hlen,
+      Cstruct.sub keys 128 klen,
+      Cstruct.sub keys 192 hlen )
+  in
+  let keys =
+    match cipher with
+    | `AES_256_CBC ->
+        let hmac_len = Mirage_crypto.Hash.digest_size hmac_algorithm in
+        let my_key, my_hmac, their_key, their_hmac =
+          maybe_swap (extract 32 hmac_len)
+        in
+        AES_CBC
+          {
+            my_key = Mirage_crypto.Cipher_block.AES.CBC.of_secret my_key;
+            my_hmac;
+            their_key = Mirage_crypto.Cipher_block.AES.CBC.of_secret their_key;
+            their_hmac;
+          }
+    | `AES_128_GCM ->
+        let my_key, my_implicit_iv, their_key, their_implicit_iv =
+          maybe_swap (extract 16 (Packet.aead_nonce - Packet.packet_id_len))
+        in
+        AES_GCM
+          {
+            my_key = Mirage_crypto.Cipher_block.AES.GCM.of_secret my_key;
+            my_implicit_iv;
+            their_key = Mirage_crypto.Cipher_block.AES.GCM.of_secret their_key;
+            their_implicit_iv;
+          }
+    | `AES_256_GCM ->
+        let my_key, my_implicit_iv, their_key, their_implicit_iv =
+          maybe_swap (extract 32 (Packet.aead_nonce - Packet.packet_id_len))
+        in
+        AES_GCM
+          {
+            my_key = Mirage_crypto.Cipher_block.AES.GCM.of_secret my_key;
+            my_implicit_iv;
+            their_key = Mirage_crypto.Cipher_block.AES.GCM.of_secret their_key;
+            their_implicit_iv;
+          }
+    | `CHACHA20_POLY1305 ->
+        let my_key, my_implicit_iv, their_key, their_implicit_iv =
+          maybe_swap (extract 32 (Packet.aead_nonce - Packet.packet_id_len))
+        in
+        CHACHA20_POLY1305
+          {
+            my_key = Mirage_crypto.Chacha20.of_secret my_key;
+            my_implicit_iv;
+            their_key = Mirage_crypto.Chacha20.of_secret their_key;
+            their_implicit_iv;
+          }
+  in
+  { my_packet_id = 1l; their_packet_id = 1l; keys }
```

Comment:
the reason behind the KDF change is that only at the point where the server send the push_reply we know which cipher to use for the data channel. this means we need to delay the keying until we are done with the push_reply.

the nice thing is that now these function are separate and not in the result monad :)